notify/sms/conf/conf.go

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2022, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package conf
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/megaease/easeprobe/notify/base"
+)
+
+// Provider Interface
+type Provider interface {
+	Kind() string
+	Notify(string, string) error
+}
+
+// ProviderType is the sms provider
+type ProviderType int
+
+// The Provider Type of different sms provider
+const (
+	Unknown ProviderType = iota
+	Yunpian
+	Twilio
+	Nexmo
+)
+
+// ProviderMap is the map of [provider, name]
+var ProviderMap = map[ProviderType]string{
+	Yunpian: "yunpian",
+	Twilio:  "twilio",
+	Nexmo:   "nexmo",
+	Unknown: "unknown",
+}
+
+// Options implements the configuration for native client
+type Options struct {
+	base.DefaultNotify `yaml:",inline"`
+
+	ProviderType ProviderType `yaml:"provider"`
+	Mobile       string       `yaml:"mobile"`
+	From         string       `yaml:"from"`
+	ApiKey       string       `yaml:"apikey"`
+	ApiSecret    string       `yaml:"api_secret"`
```

Comment:
looks `ApiSecret` is only used by Twilio， I think you can move this to Twilio struct. ;-)

Reply:
> looks `ApiSecret` is only used by Twilio， I think you can move this to Twilio struct. ;-)

I have a problem here, because I don't know how to construct `ApiSecret` in Twilio struct.

Reply:
And `ApiSecret` is using by Nexmo, `AUTH_TOKEN` is using by Twilio. Maybe we can just using key and secret to abstract them.

```sh
curl -X "POST" "https://rest.nexmo.com/sms/json" \
  -d "from=$VONAGE_BRAND_NAME" \
  -d "text=A text message sent using the Vonage SMS API" \
  -d "to=$TO_NUMBER" \
  -d "api_key=$VONAGE_API_KEY" \
  -d "api_secret=$VONAGE_API_SECRET"
```

```sh
curl -X POST https://api.twilio.com/2010-04-01/Accounts/$TWILIO_ACCOUNT_SID/Messages.json \
--data-urlencode "Body=Hi there" \
--data-urlencode "From=+15017122661" \
--data-urlencode "To=+15558675310" \
-u $TWILIO_ACCOUNT_SID:$TWILIO_AUTH_TOKEN
```

Reply:
there are two kinds of construction:

```yaml
sms:
-  name "xxxx"
    driver: "twilio"
    ....
- name "yyy"
   driver: "yunpian"
   ...
```

and 

```yaml
sms:
    twilio:
    - name xxx
       ....
     yunpian:
     - name yyyy
        ...
```
- For the first one, you have to have a common struct that includes all of the fields of all of the SMS providers.
- For the second one, you can embed the common struct into all of the providers' structs. 

Both of ways are fine to me, you can make your own decision.

Reply:
But if we go the 2nd way, we have to refractory the `AllNotifiers()` function in  `conf/conf.go`, because we need to find the notification structure recursively... this probably is a bit complicated. 

So, may you can stick to your current way, that would be fewer code changes.